### PR TITLE
[Platform API][pytest] Add basic tests for PSU class

### PR DESCRIPTION
**Pull request by sujinmkang**
Add a basic test suite for the Psu class of the new platform API, utilizing the platform API HTTP server, including the following functions:
```
test_fans
test_power
test_temperature
test_led
```

Example test results summary:
```
========================================================================== short test summary info ===========================================================================
FAILED platform_tests/api/test_psu.py::TestPsuApi::test_power - Failed: Failed to retrieve the voltage threshold values of PSU 0
FAILED platform_tests/api/test_psu.py::TestPsuApi::test_temperature - Failed: Failed to retrieve temperature of PSU 0
FAILED platform_tests/api/test_psu.py::TestPsuApi::test_led - Failed: Failed to set status_led of PSU 0
==================================================================== 3 failed, 1 passed in 46.80 seconds =====================================================================
```
This result could be returned when these apis are not implemented yet in the platform api on DUT.
- get_voltage_high_threshold and get_voltage_low_threshold for test_power.
- get_temperature for test_temperature.
- set_status_led for test_led.

### CODE REVIEW

File: tests/platform_tests/api/test_psu.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    pytest.mark.sanity_check(skip_sanity=True),
+    pytest.mark.disable_loganalyzer  # disable automatic loganalyzer
+]
+
+STATUS_LED_COLOR_GREEN = "green"
+STATUS_LED_COLOR_AMBER = "amber"
+STATUS_LED_COLOR_RED = "red"
+STATUS_LED_COLOR_OFF = "off"
+
+class TestPSUAPI(object):
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:23]: Suggest to change class name to PascalCase, e.g., `TestPsuApi`. I intend to change the existing chassis and watchdog classes to match this, also.

File: tests/platform_tests/api/test_psu.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+        for psu_id in range(num_psus):
+            psu = chassis.get_psu(platform_api_conn, psu_id)
+            voltage = psu.get_voltage(platform_api_conn)
+            current = psu.get_current(platform_api_conn)
+            power = psu.get_power(platform_api_conn)
+            pytest_assert(abs(power - (voltage*current)) > power*0.1, "Power is too off compared to the voltage and current reading")
+
+            powergood_status = psu.get_powergood_status(platform_api_conn)
+            pytest_assert(powergood_status is not True, "PSU Powergood status test returns failed")
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:64]: I suggest performing two checks, one to ensure the API is implemented, and if so, one to check the sanity of the value. Also please indicate the index of the PSU in the error message. E.g.,

```
pytest_assert(powergood_status is not None, "Failed to retrieve operational status of PSU {}".format(psu_id))
pytest_assert(powergood_status is not True, "PSU {} is not operational".format(psu_id))
```

File: tests/platform_tests/api/test_psu.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+            try:
+                high_threshold = psu.get_voltage_high_threshold(platform_api_conn)
+                low_threshold = psu.get_voltage_low_threshold(platform_api_conn)
+            except NotImplementedError:
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:69]: You will never catch `NotImplementedError` in these tests, because you are not calling the API directly, but rather the result is being sent from the HTTP server. I have made a fix [here](https://github.com/Azure/sonic-mgmt/pull/1822) to ensure that if the HTTP server catches a `NotImplementedError`, it will respond with None. This is why I suggest checking each value against None first, as I mentioned in another comment above.

File: tests/platform_tests/api/test_psu.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        try:
+            num_psus = int(chassis.get_num_psus(platform_api_conn))
+        except:
+            pytest_assert("num_psus is not integer!")
+
+        for psu_id in range(num_psus):
+            psu = chassis.get_psu(platform_api_conn, psu_id)
+            try:
+                temperature = psu.get_temperature(platform_api_conn)
+            except NotImplementedError:
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:87]: See my comment about `NotImplementedError` above.

File: tests/platform_tests/api/test_psu.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                logger.warning('get_temperature is not implemented yet') 
+            else:
+                try:
+                    temp_threshold = psu.get_temperature_high_threshold(platform_api_conn)
+                except NotImplementedError:
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:92]: See my comment about `NotImplementedError` above.

File: tests/platform_tests/api/test_psu.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        except:
+            pytest_assert("num_psus is not integer!")
+
+        for psu_id in range(num_psus):
+            psu = chassis.get_psu(platform_api_conn, psu_id)
+
+            for color in LED_COLOR_LIST:
+                try:
+                    result = psu.set_status_led(platform_api_conn, color)
+                except NotImplementedError:
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:117]: See my comment about `NotImplementedError` above.

File: tests/platform_tests/api/test_psu.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            num_psus = int(chassis.get_num_psus(platform_api_conn))
+        except:
+            pytest_assert("num_psus is not integer!")
+
+        for psu_id in range(num_psus):
+            psu = chassis.get_psu(platform_api_conn, psu_id)
+            voltage = psu.get_voltage(platform_api_conn)
+            current = psu.get_current(platform_api_conn)
+            power = psu.get_power(platform_api_conn)
+            pytest_assert(abs(power - (voltage*current)) > power*0.1, "Power is too off compared to the voltage and current reading")
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:61]: Suggest rewording error message and adding values and PSU index into it. E.g.,

```
"PSU {} readings do not make sense (power: {}, voltage: {}, current: {})".format(psu_id, power, voltage, current)
```

**Comment by sujinmkang** [comment]
retest vsimage please

File: tests/platform_tests/api/test_psu.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+from common.helpers.assertions import pytest_assert
+from common.helpers.platform_api import psu
+from common.helpers.platform_api import chassis
+
+logger = logging.getLogger(__name__)
+
+pytestmark = [
+    pytest.mark.sanity_check(skip_sanity=True),
+    pytest.mark.disable_loganalyzer  # disable automatic loganalyzer
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:15]: Recently there was a new 'topology' marker added. Please mark these tests to run on 'any' topology. See https://github.com/Azure/sonic-mgmt/blob/master/tests/platform_tests/api/test_chassis.py#L14 for an example.

File: tests/platform_tests/api/test_psu.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            self.expect(abs(power - (voltage*current)) < power*0.1, "PSU {} reading does not make sense \
+                (power:{}, voltage:{}, current:{})".format(psu_id, power, voltage, current))
+
+            powergood_status = psu.get_powergood_status(platform_api_conn, psu_id)
+            if self.expect(powergood_status is not None, "Failed to retrieve operational status of PSU {}".format(psu_id)):
+                self.expect(powergood_status is True, "PSU {} is not operational".format(psu_id))
+
+            high_threshold = psu.get_voltage_high_threshold(platform_api_conn, psu_id)
+            low_threshold = psu.get_voltage_low_threshold(platform_api_conn, psu_id)
+            if self.expect(high_threshold is not None and low_threshold is not None, "Failed to retrieve the voltage threshold values of PSU {}".format(psu_id)):
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:73]: Suggest to split this into separate assertions for high_threshold and low_threshold so that we can get more clear error messages of which one (or both) failed. Maybe something like:

```
high_threshold = psu.get_voltage_high_threshold(platform_api_conn, psu_id)
if self.expect(high_threshold is not None, "Failed to retrieve the high voltage threshold value of PSU {}".format(psu_id)):
    self.expect(<check that high_threshold value is of proper expected type (float?) here> ...)

low_threshold = psu.get_voltage_low_threshold(platform_api_conn, psu_id)
if self.expect(low_threshold is not None, "Failed to retrieve the low voltage threshold value of PSU {}".format(psu_id)):
    self.expect(<check that low_threshold value is of proper expected type (float?) here> ...)

if high_threshold and low_threshold:
    self.expect(voltage < high_threshold and voltage > low_threshold, "Voltage {} of PSU {} is not in between {} and {}".format(voltage, psu_id, low_threshold, high_threshold))
```

File: tests/platform_tests/api/test_psu.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                self.expect(fan and fan == fan_list[i], "Fan {} of PSU {} is incorrect".format(i, psu_id))
+        self.assert_expectations()
+
+
+    def test_power(self, duthost, localhost, platform_api_conn):
+        ''' PSU power test '''
+        for psu_id in range(self.num_psus):
+            voltage = psu.get_voltage(platform_api_conn, psu_id)
+            current = psu.get_current(platform_api_conn, psu_id)
+            power = psu.get_power(platform_api_conn, psu_id)
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:63]: Suggest to check that these values are not None and are of the expected type.

File: tests/platform_tests/api/test_psu.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            if self.expect(high_threshold is not None and low_threshold is not None, "Failed to retrieve the voltage threshold values of PSU {}".format(psu_id)):
+                self.expect(voltage < high_threshold and voltage > low_threshold, "Voltage {} of PSU {} is not in between {} and {}".format(voltage, psu_id, low_threshold, high_threshold))
+        self.assert_expectations()
+
+
+    def test_temperature(self, duthost, localhost, platform_api_conn):
+        ''' PSU temperature test '''
+        for psu_id in range(self.num_psus):
+            temperature = psu.get_temperature(platform_api_conn, psu_id)
+            self.expect(temperature is not None, "Failed to retrieve temperature of PSU {}".format(psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:82]: Suggest to also check that value if of the expected type.

Comment (sujinmkang) [reply tests/platform_tests/api/test_psu.py:82]: @jleveque the value validation check is down below after reading the temperature threshold.

Comment (jleveque) [reply tests/platform_tests/api/test_psu.py:82]: There you're comparing `temperature < temp_threshold`. However, we should also ensure we get the proper type, as Python could allow the vendor to return an int instead of a float, etc, which the comparison alone won't catch. I have made similar checks in the SFP tests (example: https://github.com/Azure/sonic-mgmt/blob/master/tests/platform_tests/api/test_sfp.py#L223).

I think we should always make three checks on return values;
1. Value is not None (i.e., vendor has implemented the API call)
2. Value is of appropriate type (float, int, bool, str, etc.)
3. Value is acceptable (within range, tolerance, etc.), for values we have data to compare against

File: tests/platform_tests/api/test_psu.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+                fan = psu.get_fan(platform_api_conn, psu_id, i)
+                self.expect(fan and fan == fan_list[i], "Fan {} of PSU {} is incorrect".format(i, psu_id))
+        self.assert_expectations()
+
+
+    def test_power(self, duthost, localhost, platform_api_conn):
+        ''' PSU power test '''
+        for psu_id in range(self.num_psus):
+            voltage = psu.get_voltage(platform_api_conn, psu_id)
+            self.expect(voltage is not None, "Failed to retrieve voltage of PSU {}".format(psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:62]: Also check that type is float like you did for temperature.

File: tests/platform_tests/api/test_psu.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+            current = psu.get_current(platform_api_conn, psu_id)
+            self.expect(current is not None, "Failed to retrieve current of PSU {}".format(psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:64]: Also check that type is float like you did for temperature.

File: tests/platform_tests/api/test_psu.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+            power = psu.get_power(platform_api_conn, psu_id)
+            self.expect(current is not None, "Failed to retrieve current of PSU {}".format(psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:66]: Also check that type is float like you did for temperature.

File: tests/platform_tests/api/test_psu.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+            if current is not None and voltage is not None and power is not None:
+                self.expect(abs(power - (voltage*current)) < power*0.1, "PSU {} reading does not make sense \
+                    (power:{}, voltage:{}, current:{})".format(psu_id, power, voltage, current))
+
+            powergood_status = psu.get_powergood_status(platform_api_conn, psu_id)
+            if self.expect(powergood_status is not None, "Failed to retrieve operational status of PSU {}".format(psu_id)):
+                self.expect(powergood_status is True, "PSU {} is not operational".format(psu_id))
+
+            high_threshold = psu.get_voltage_high_threshold(platform_api_conn, psu_id)
+            self.expect(high_threshold is not None, "Failed to retrieve the high voltage threshold of PSU {}".format(psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:76]: Also check that type is float like you did for temperature.

File: tests/platform_tests/api/test_psu.py [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+            low_threshold = psu.get_voltage_low_threshold(platform_api_conn, psu_id)
+            self.expect(low_threshold is not None, "Failed to retrieve the low voltage threshold of PSU {}".format(psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:78]: Also check that type is float like you did for temperature.

File: tests/platform_tests/api/test_psu.py [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+    def test_temperature(self, duthost, localhost, platform_api_conn):
+        ''' PSU temperature test '''
+        for psu_id in range(self.num_psus):
+            temperature = psu.get_temperature(platform_api_conn, psu_id)
+            if self.expect(temperature is not None, "Failed to retrieve temperature of PSU {}".format(psu_id)):
+                self.expect(isinstance(temperature, float), "PSU {} temperature appears incorrect".format(psu_id))
+
+            temp_threshold = psu.get_temperature_high_threshold(platform_api_conn, psu_id)
+            if self.expect(temp_threshold is not None, "Failed to retrieve temperature threshold of PSU {}".format(psu_id)):
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:92]: Also check that type is float like you did for temperature.

File: tests/platform_tests/api/test_psu.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+            STATUS_LED_COLOR_OFF
+        ]
+
+        for psu_id in range(self.num_psus):
+            for color in LED_COLOR_LIST:
+                result = psu.set_status_led(platform_api_conn, psu_id, color)
+                if self.expect(result is not None, "Failed to perform set_status_led of PSU {}".format(psu_id)):
+                    self.expect(result is True, "Failed to set status_led of PSU {}".format(psu_id))
+
+                color_status = psu.get_status_led(platform_api_conn, psu_id)
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:118]: Check that `color_status` is not `None` and that is is of type `str`.

File: tests/platform_tests/api/test_psu.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+                if self.expect(color_status is not None, "Failed to retrieve status_led of PSU {}".format(psu_id)):
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:120]: Suggest to reword message: `Failed to retrieve status LED color of PSU {}`

File: tests/platform_tests/api/test_psu.py [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+                    if self.expect(isinstance(color_status, str), "PSU {} status led color appears incorrect".format(psu_id)):
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:121]: s/led/LED/

File: tests/platform_tests/api/test_psu.py [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+                        self.expect(color == color_status, "Retrived the status_led {} not {} from PSU {}".format(color_status, color, psu_id))
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:122]: Suggest to reword message: `"Status LED color incorrect (expected: {}, actual: {}) for PSU {}".format(color, color_status, psu_id))`. Also suggest changing variable names: `color` -> `color_target`, `color_status` -> `color_actual`.

Comment (sujinmkang) [reply tests/platform_tests/api/test_psu.py:122]: I can change color_status to color_actual but will keep the color as it is. it is still readable enough, I think. :)

Comment (jleveque) [reply tests/platform_tests/api/test_psu.py]: OK. Please also update the message per my suggestion.

**Comment by sujinmkang** [comment]
retest this please

File: tests/platform_tests/api/test_psu.py (hunk 21 above)
Comment (jleveque) [reply tests/platform_tests/api/test_psu.py]: FYI, the message is still not updated.

File: tests/platform_tests/api/test_psu.py (hunk 19 above)
Comment (jleveque) [reply tests/platform_tests/api/test_psu.py]: Please update this message as suggested.

Comment (jleveque) [reply tests/platform_tests/api/test_psu.py]: Please update this message as suggested.

File: tests/platform_tests/api/test_psu.py [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+            STATUS_LED_COLOR_GREEN,
+            STATUS_LED_COLOR_AMBER,
+            STATUS_LED_COLOR_RED,
+            STATUS_LED_COLOR_OFF
+        ]
+
+        for psu_id in range(self.num_psus):
+            for color in LED_COLOR_LIST:
+                result = psu.set_status_led(platform_api_conn, psu_id, color)
+                if self.expect(result is not None, "Failed to perform set_status_led of PSU {}".format(psu_id)):
```
Comment (jleveque) [inline tests/platform_tests/api/test_psu.py:116]: Suggest to update message as follows:
```
"Failed to set status LED of PSU {} to {}".format(psu_id, color)
```
